Improve robustness of return mapping iterations

bwspenc opened the pull request:
ref #6873
- Adds new base class that provides return mapping capability.  This is
  used currently by solid_mechanics models, but it will be straightforward
  to move tensor_mechanics models over to it.
- Add line search and bisection to drastically improve the robustness of
  return mapping iterations.
- Make all models compute residuals in strain increment units for consistency.
- Use the elastic strain magnitude as the reference quantity for relative
  convergence (rather than the initial residual, which can be extremely high)
- Set the default convergence tolerances to much tighter values
- Remove the option to output iterations on failure.  Instead, always re-run
  the model with output turned on if it fails to converge.
- Add _qp to the ConstitutiveModel base class to be more consistent with
  tensor_mechanics, and make it easier to base the tensor_mechanics models on
  this.
- Add a legacy_return_mapping option to give access to the previous return
  mapping procedure and tolerances as needed.  This is intended only to be
  temporarily available for compatibility with existing models.

bwspenc committed:
Improve robustness of return mapping iterations ref #6873

- Adds new base class that provides return mapping capability.  This is
  used currently by solid_mechanics models, but it will be straightforward
  to move tensor_mechanics models over to it.
- Add line search and bisection to drastically improve the robustness of
  return mapping iterations.
- Make all models compute residuals in strain increment units for consistency.
- Use the elastic strain magnitude as the reference quantity for relative
  convergence (rather than the initial residual, which can be extremely high)
- Set the default convergence tolerances to much tighter values
- Remove the option to output iterations on failure.  Instead, always re-run
  the model with output turned on if it fails to converge.
- Add _qp to the ConstitutiveModel base class to be more consistent with
  tensor_mechanics, and make it easier to base the tensor_mechanics models on
  this.
- Add a legacy_return_mapping option to give access to the previous return
  mapping procedure and tolerances as needed.  This is intended only to be
  temporarily available for compatibility with existing models.

bwspenc commented:
@sapitts @acasagran @dschwen Would one or more of you mind taking a look at this?

This does break BISON due to API incompatibilites, and I have a patch ready.  With the exception of the couple of cases we discussed that have issues unrelated to this, all of the BISON assessment cases run with this fix, and with only default convergence settings, which are far tighter than the current tolerances.

moosebuild commented:
Job [Documentation](https://www.moosebuild.org/job/99090/) on 5b6843e wanted to post the following:

View the site [here](http://mooseframework.org/docs/PRs/9445/site)

dschwen commented on modules/solid_mechanics/include/materials/NewtonIterationSingleVariable.h:
```diff
@@ -0,0 +1,214 @@
+ * iterations on a single variable
+ */
+
+class NewtonIterationSingleVariable;
```
This name is a bit too general for something that requires a _trial stress_ in my opinion.

bwspenc replied:
I had the same thought.  I had the idea originally that this would be something used for general single-variable Newton solves, and maybe it's possible to re-working it to be more general in the future.  I propose that for now, we keep this as a return mapping solver (which it appears can already be shared at least 4 different places), and then maybe revisit the idea of making it more general in the future.  I renamed it to SingleVariableReturnMappingSolution to reflect the fact that it is a return mapping solver. Since this is a base class, changing the name won't affect any input files.

WilkAndy commented:
In TensorMechanics i've created MultiParameterPlasticityStressUpdate (MPPSU) that does the return-map using a line search (between lines 283 and 317 currently).  I think it's all pretty standard except that i check for precision loss because it has turned out to be quite important in plasticity models.  I hope that when you do the equivalent of this PR in TensorMechanics you can either use the code in MPPSU or make MPPSU use your code.  I've spoken to @fdkong about whether MOOSE could include its own Newton-Raphson+line-search because we shouldn't really have multiple bits of code doing the same thing.

sapitts commented:
@WilkAndy I quite like this idea of a common Newton Rhapson + line search.  @jiangwen84 and I  also have another line search method in our crystal plasticity codes which would further strengthen the argument for a common utility like the one you suggested.

sapitts commented on modules/solid_mechanics/include/materials/ReturnMappingModel.h:
```diff
@@ -8,61 +8,79 @@
- * One or more constitutive models coupled together.
+ * Base class for models that perform return mapping iterations to compute stress.  This
+ * is for a single model.  This class can be used directly for a single model, or combined
+ * with other models.
```
How combined with other models?  Will the multiple models through the picard iterations in tensor mechanics work in a similar fashion to how they do now?

bwspenc replied:
I revised the comment -- hopefully it makes more sense now!

sapitts commented on modules/solid_mechanics/src/materials/NewtonIterationSingleVariable.C:
```diff
@@ -0,0 +1,443 @@
+                                                  const ConsoleStream & console)
+{
+  std::stringstream iter_output;
+  std::stringstream * iter_output_ptr = (_output_iteration_info ? &iter_output : nullptr);
```
Is stringstream the right "moosey" thing to use here?

bwspenc replied:
I discussed this with @dschwen, and came to the conclusion that there was nothing un-moosey about using a stringstream.  I think it's more natural here.

sapitts commented on modules/solid_mechanics/src/materials/NewtonIterationSingleVariable.C:
```diff
@@ -0,0 +1,443 @@
+                            iter_output);
+
+    residual = computeResidualInternal(effective_trial_stress, scalar, dt);
+    reference_residual = computeReferenceResidual(effective_trial_stress, scalar);
```
Is this reference residual related to the reference residual in Problem that we use in the assessment cases?

bwspenc replied:
Similar idea.  It's the quantity that we compare the residual to for relative convergence.

sapitts commented on modules/solid_mechanics/src/materials/NewtonIterationSingleVariable.C:
```diff
@@ -0,0 +1,443 @@
+    _relative_tolerance(parameters.get<Real>("relative_tolerance")),
+    _absolute_tolerance(parameters.get<Real>("absolute_tolerance")),
+    _line_search(true),
+    _bracket_solution(true),
```
What does this variable do?

bwspenc replied:
See my Doxygen comment in the header file.

sapitts commented on modules/solid_mechanics/src/materials/NewtonIterationSingleVariable.C:
```diff
@@ -0,0 +1,443 @@
+            Real frac = 0.5;
+            scalar_increment =
+                (1.0 - frac) * scalar_lower_bound + frac * scalar_upper_bound - scalar_old;
+            // scalar_increment = (scalar_upper_bound + scalar_lower_bound) / 2.0 - scalar_old;
```
Remove if unneded

bwspenc replied:
Done

sapitts commented on modules/solid_mechanics/src/materials/NewtonIterationSingleVariable.C:
```diff
@@ -0,0 +1,443 @@
+  InputParameters params = emptyInputParameters();
+
+  // Newton iteration control parameters
+  params.addParam<unsigned int>("max_its", 200, "Maximum number of Newton iterations");
```
While we are breaking BISON, could we change the variable name from _max_its to _max_iterations?

bwspenc replied:
There were enough cases in the BISON assessment cases that failed with a lower number that I decided to bump it up.  Those are usually cases that have crazy predictor stresses on the order of 1e100, and they actually converge if you give them 200 iterations.  The old method computed the residual relative to the initial residual, so it could happily accept a solution with a residual on the order of 1e100.

bwspenc replied:
I'd be in favor of deprecating max_its.  Maybe we could do that in a separate commit so we don't break more than we have to all at once.

sapitts commented on modules/solid_mechanics/src/materials/PowerLawCreepModel.C:
```diff
@@ -55,33 +54,33 @@ PowerLawCreepModel::computeStressInitialize(unsigned qp,
+  return -3.0 * _coefficient * _shear_modulus * _n_exponent *
+             std::pow(effectiveTrialStress - 3.0 * _shear_modulus * scalar, _n_exponent - 1.0) *
+             _exponential * _expTime * _dt -
+         1.0;
```
This looks like a clang format residual; can the 1.0 be moved up to the previous line for code readability without making clang format unhappy?

bwspenc replied:
I don't really think so, unfortunately!

sapitts reviewed (commented):
Looks good, and I'm excited to get this capability in both solid and tensor mechanics.  Mainly a number of documentation and variable naming comments, but nothing that would prevent the code from being merged.

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/99394/) on 5b8d938 wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format bcb391f274217a7bfba326784f990568ca55f372`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/9445/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/9445/style.patch | git apply -v`

bwspenc commented on modules/solid_mechanics/include/materials/NewtonIterationSingleVariable.h:
```diff
@@ -0,0 +1,214 @@
+   * Perform the return mapping iterations
+   * @param effective_trial_stress Effective trial stress
+   * @param scalar                 Inelastic strain increment magnitude being solved for
+   * @param console                Console output
```
I believe this is how we're supposed to document the code for Doxygen.  I've personally never been very convinced of the benefits of Doxygen, but I'm pretty sure that's the coding standard.

bwspenc commented on modules/solid_mechanics/include/materials/NewtonIterationSingleVariable.h:
```diff
@@ -0,0 +1,214 @@
+  /**
+   * Compute the maximum permissible value of the scalar.  For some models, the magnitude
+   * of this may be known.
+   * @param effective_trial_stress Effective trial stress
```
This is exactly why I don't think the Doxygen comments are particularly useful, but we're supposed to document every single parameter, even if they have self-explanatory names.

bwspenc commented on modules/solid_mechanics/src/materials/CLSHPlasticModel.C:
```diff
@@ -58,46 +54,46 @@ CLSHPlasticModel::computeStressInitialize(unsigned qp,
     _xphidp = -3. * _shear_modulus * _c_alpha * _c_beta * std::cosh(xflow);
     _xphir = -_c_alpha * _c_beta * std::cosh(xflow);
-    residual = xphi - scalar / _dt;
+    residual = xphi * _dt - scalar;
```
Yes

bwspenc replied:
Yes

bwspenc commented on modules/solid_mechanics/include/materials/ReturnMappingModel.h:
```diff
@@ -8,61 +8,79 @@
   Real _effective_strain_increment;
 
+  /// 3 * shear modulus
+  Real _three_g;
```
Sure, I can make that change.  I think the standard is that we're supposed to Doxygen coment all of these variables, even if they have self-explanatory names, which gives more opportunities for comments that are lies.  Did I mention that I see no value in Doxygen?

bwspenc commented on modules/solid_mechanics/include/materials/ReturnMappingModel.h:
```diff
@@ -8,61 +8,79 @@
  */
 
-class ReturnMappingModel : public ConstitutiveModel
+class ReturnMappingModel : public ConstitutiveModel, public NewtonIterationSingleVariable
```
We will follow this same pattern.  RadialReturnStressUpdate will inherit from both StressUpdateBase and NewtonIterationSingleVariable (which I have renamed now), and we'll get rid of all of the Newton iteration code in RadialReturnStressUpdate.  This structure will allow us to accomplish exactly what Andy is suggesting.

bwspenc commented on modules/solid_mechanics/src/materials/NewtonIterationSingleVariable.C:
```diff
@@ -0,0 +1,443 @@
+  params.addParam<Real>(
+      "relative_tolerance", 1e-8, "Relative convergence tolerance for Newton iteration");
+  params.addParam<Real>(
+      "absolute_tolerance", 1e-11, "Absolute convergence tolerance for Newton iteration");
```
I agree with your point, but this is a general base class that isn't intended to prescribe the specifics of that.  That's something determined by the models that derive from this.

bwspenc replied:
Maybe we should redefine this parameter in the derived class for a descriptive string that makes sense for that class?

bwspenc commented on modules/solid_mechanics/src/materials/NewtonIterationSingleVariable.C:
```diff
@@ -0,0 +1,443 @@
+  if (parameters.get<bool>("legacy_return_mapping") == true)
+  {
+    if (!parameters.isParamSetByUser("relative_tolerance"))
+      _relative_tolerance = 1.e-5;
```
In general, the default was 1e-5, but there is a BISON model that actually set it looser.  I have retained that behavior in my BISON code that accompanies this.

bwspenc commented:
@WilkAndy I wasn't aware of your other version of return mapping iterations.  We should be able to make that derive from this class as well (probably after pulling in a few modifications from your code into it).  I am definitely doing this with the goal of removing duplicated code in mind.  I've put quite a bit of time into making sure that the set of ~80 long-running BISON test cases works with the current code, so I propose we keep this code as-is for now, and then look into the differences between it and your code in the next pass through this.

WilkAndy commented:
Sounds good to me!  One thing that i haven't implemented is a region of "trust" in the line search.  That is, the NewtonRaphson might send the variables into an unphysical region where we don't want to go, even if the residual is lower there, but the line search (or something else) should be able to prevent this

bwspenc commented on modules/solid_mechanics/include/materials/CombinedCreepPlasticity.h:
```diff
@@ -23,14 +23,12 @@ class CombinedCreepPlasticity : public ConstitutiveModel
                              SymmTensor & stress_new);
 
   virtual bool modifyStrainIncrement(const Elem & current_elem,
-                                     unsigned qp,
```
This still isn't completely foolproof because the model that call it has to set the qp before it calls it, but that only happens in a couple of places in the code, and matches the tensor_mechanics pattern.  In the models themselves, you just always use _qp now, which I think is easier.

bwspenc commented:
@sapitts Thanks for reviewing this.  I think I addressed all of your issues, and you probably saw that I submitted a corresponding patch to BISON.  If you could review that, I'd appreciate it.  Since getting this pushed through will involve breaking dependent codes, it's probably best if we do this when I get back in the office, probably Monday.  @dschwen, since you're the one who will probably need to merge this, if you could review it so that I can address any issues and have it ready to go Monday, I'd appreciate it.

dschwen commented on modules/tensor_mechanics/include/materials/SingleVariableReturnMappingSolution.h:
```diff
@@ -0,0 +1,191 @@
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+#ifndef NEWTONITERATIONSINGLEVARIABLE_H
+#define NEWTONITERATIONSINGLEVARIABLE_H
```
the include guard needs to be fixed

bwspenc replied:
done

dschwen commented on modules/tensor_mechanics/include/materials/SingleVariableReturnMappingSolution.h:
```diff
@@ -0,0 +1,191 @@
+/**
+ * Base class that provides capability for Newton return mapping
+ * iterations on a single variable
+ */
```
can you move that to just above the actual class declaration

bwspenc replied:
done
